src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -692,15 +692,15 @@ private void writeRecord(SinkRecord record) {
   }
 
   private void closeTempFile(String encodedPartition) {
-    if (writers.containsKey(encodedPartition)) {
-      io.confluent.connect.storage.format.RecordWriter writer = writers.get(encodedPartition);
+    io.confluent.connect.storage.format.RecordWriter writer = writers.remove(encodedPartition);
+    if (writer != null) {
```

Comment:
This is a slightly more efficient approach, and more tolerant should there not be a writer for an encoded partition.

Reply:
Something occurred to me: by removing the writer *before* attempting to close it, we will lose track of the (un)closed writer should there be an exception during `close()` on the next few lines.

In fact, I think if we don't do this, then if an exception happens during `close()`, the existing code would not remove the `RecordWriter`, which means that `getWriter(String encodedPartition)` would return this improperly closed writer, and the `TopicPartitionWriter` would continue to use the writer even though it is an unexpected state. This is in contrast to the contract for `Closeable`, which is that once `close()` is called the resource not be used anymore. I think we see evidence of this: [a particular NPE in the ParquetWriter](https://issues.apache.org/jira/browse/PARQUET-544) is often attributed to `close()` being called more than once, and the `TopicPartitionWriter` would do this if the first `close()` failed and the `RecordWriter` were left in this `writers` map (as is the case in the current code).

So, I think that we should remove the `RecordWriter` regardless of whether or not we're able to successfully `close()` the writer -- as long as we at least call `close()`, which the current PR does. The other benefit is that when `close()` does fail and the task ultimately retries the topic partition, the `getWriter(...) method will not see in the `writers` map the writer for which `close()` failed and will create a new one and put it into the `writers` map. This is the behavior we want, as subsequent attempts to retry writing records to the topic partition will be performed with a writer re-acquired after the `close()` failure and hopefully when HDFS has healed itself.

I did add another commit that documents this reasoning in a comment just before this line of code.